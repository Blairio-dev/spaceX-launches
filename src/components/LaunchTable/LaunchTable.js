import React from "react";
import PropTypes from "prop-types";
import styled from "@emotion/styled";
import moment from "moment";
import { colours } from "../../assets/tokens";
import { Button, Inline } from "../../components";
import { Select } from "../Select/Select";
import { sortLaunches } from "../../assets/functions";

const StyledControls = styled("div")`
  margin-bottom: 13px;

  select {
    margin-right: 9px;
  }
`;

const StyledRow = styled("tr")`
  align-items: center;
  border-radius: 10px;
  box-shadow: 0px 2px 15px 0px hsl(0, 0%, 0%, 25%);
  color: ${colours.text};
  display: flex;
  min-height: 80px;
  justify-content: space-between;
  margin-bottom: 6px;
`;

const StyledFlightNumber = styled("td")`
  font-family: "Brandon Grotesque Medium";
  font-size: 40px;
  font-weight: 500;
  margin-left: 32px;
  margin-right: 24px;
  width: 76px;
`;

const StyledMissionName = styled("td")`
  font-family: "Brandon Grotesque Medium";
  font-size: 32px;
  font-weight: 500;
  width: 100%;
`;

const StyledLaunchDate = styled("div")`
  display: flex;
  font-family: "Brandon Grotesque Regular";
  font-size: 16px;
  justify-content: inherit;
  margin-bottom: 8px;
  white-space: nowrap;
`;

const StyledRocketName = styled("div")`
  display: flex;
  font-family: "Brandon Grotesque Bold";
  font-size: 24px;
  justify-content: inherit;
`;

const StackedCell = styled("td")`
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  margin-left: 21px;
  margin-right: 16px;
`;

const StyledTable = styled("table")`
  border-spacing: 0;
  width: 100%;

  td {
    padding: 0;
  }
`;

const LaunchTable = ({
  filterOnChange,
  isAscending,
  launchData,
  launchYear,
  sortOnClick,
}) => {
  const launchYearOptions = [
    ...new Set(launchData.map((launch) => launch.launch_year)),
  ];
  const filteredLaunches = sortLaunches(launchData, launchYear, isAscending);
  return (
    <div>
      <StyledControls>
        <Inline.RightAligned
          nodes={[
            <Select
              labelText="Filter by Year"
              key="launch-year-filter-select"
              id="launch-year-filter-select"
              onChange={(event) => filterOnChange(event)}
              selectOptions={["Filter by Year", ...launchYearOptions]}
              selectedOption={launchYear}
            />,
            <Button.Sort
              labelText={`Sort ${isAscending ? "Descending" : "Ascending"}`}
              key="button-sort"
              onClick={sortOnClick}
            />,
          ]}
        />
      </StyledControls>
      <StyledTable>
        <tbody>
          {filteredLaunches.map((launch, index) => (
            <StyledRow key={`#${launch.id}-${launch.mission_name}-${index}`}>
              <StyledFlightNumber>{`#${launch.id}`}</StyledFlightNumber>
              <StyledMissionName>{launch.mission_name}</StyledMissionName>
              <StackedCell>
                <StyledLaunchDate>
                  {moment(launch.launch_date_utc).format("Do MMM YYYY")}
                </StyledLaunchDate>
                <StyledRocketName>{launch.rocket.rocket_name}</StyledRocketName>
              </StackedCell>
            </StyledRow>
          ))}
        </tbody>
      </StyledTable>
    </div>
  );
};

LaunchTable.propTypes = {
  /** Set to true when data should be sorted in ascending order. */
  isAscending: PropTypes.bool.isRequired,
  /** Callback function triggered on change of the filter select. */
  filterOnChange: PropTypes.func.isRequired,
  /** Launch data to populate the table. */
  launchData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      mission_name: PropTypes.string.isRequired,
      launch_date_utc: PropTypes.string.isRequired,
      launch_year: PropTypes.string.isRequired,
      rocket: PropTypes.PropTypes.shape({
        rocket_name: PropTypes.string.isRequired,
      }),
    })
  ).isRequired,
  /** Selected launch year used to filter data. */
  launchYear: PropTypes.string.isRequired,
  /** Callback function triggered on click of sort button. */
  sortOnClick: PropTypes.func.isRequired,
};

LaunchTable.defaultProps = {
  isAscending: true,
};

export { LaunchTable };
