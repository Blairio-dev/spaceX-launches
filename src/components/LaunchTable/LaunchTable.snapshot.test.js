import React from "react";
import renderer from "react-test-renderer";
import { LaunchTable } from "./LaunchTable";

const testData = [
  {
    id: "1",
    mission_name: "FalconSat",
    launch_date_utc: "2006-03-24T22:30:00.000Z",
    launch_year: "2006",
    rocket: {
      rocket_name: "Falcon 1",
    },
  },
  {
    id: "2",
    mission_name: "DemoSat",
    launch_date_utc: "2007-03-21T01:10:00.000Z",
    launch_year: "2007",
    rocket: {
      rocket_name: "Falcon 1",
    },
  },
  {
    id: "3",
    mission_name: "Trailblazer",
    launch_date_utc: "2008-08-03T03:34:00.000Z",
    launch_year: "2008",
    rocket: {
      rocket_name: "Falcon 1",
    },
  },
  {
    id: "4",
    mission_name: "RatSat",
    launch_date_utc: "2008-09-28T23:15:00.000Z",
    launch_year: "2008",
    rocket: {
      rocket_name: "Falcon 1",
    },
  },
  {
    id: "5",
    mission_name: "RazakSat",
    launch_date_utc: "2009-07-13T03:35:00.000Z",
    launch_year: "2009",
    rocket: {
      rocket_name: "Falcon 1",
    },
  },
];

const DEFAULT_PROPS = {
  filterOnChange: jest.fn(),
  launchData: testData,
  launchYear: "",
  sortOnClick: jest.fn(),
};

const FULLEST_PROPS = {
  ...DEFAULT_PROPS,
  isAscending: true,
  launchYear: "2008",
};

describe("LaunchTable Snapshots", () => {
  it("Renders with default props", () => {
    const tree = renderer.create(<LaunchTable {...DEFAULT_PROPS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Renders with fullest props", () => {
    const tree = renderer.create(<LaunchTable {...FULLEST_PROPS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
