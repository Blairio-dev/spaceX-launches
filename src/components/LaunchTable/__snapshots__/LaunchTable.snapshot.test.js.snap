// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchTable Snapshots Renders with default props 1`] = `
<div>
  <div
    className="css-m5epdm"
  >
    <div
      className="css-17w8ll9"
    >
      <div
        className="css-ovst58"
      >
        <select
          className="css-1bjzmyt"
          id="launch-year-filter-select"
          onChange={[Function]}
          value=""
        >
          <option
            value="Filter by Year"
          >
            Filter by Year
          </option>
          <option
            value="2006"
          >
            2006
          </option>
          <option
            value="2007"
          >
            2007
          </option>
          <option
            value="2008"
          >
            2008
          </option>
          <option
            value="2009"
          >
            2009
          </option>
        </select>
        <img
          alt="Chevron down"
          height="5px"
          srcSet="
			select@3x.png 3x,
			select@2x.png 2x,
			select@3x.png 1x,
		"
        />
      </div>
      <button
        className="css-4otj2c"
        onClick={[MockFunction]}
        type="button"
      >
        <span>
          Sort Descending
        </span>
        <img
          alt="Rocket launching"
          height="18px"
          srcSet="
      sort@2x.png 3x,
      sort.png 2x,
      sort.png 1x,
    "
        />
      </button>
    </div>
  </div>
  <table
    className="css-yn959"
  >
    <tbody>
      <tr
        className="css-1k3981y"
      >
        <td
          className="css-8hb6fg"
        >
          #1
        </td>
        <td
          className="css-o3rh2l"
        >
          FalconSat
        </td>
        <td
          className="css-1hyzgfm"
        >
          <div
            className="css-1wzm5rf"
          >
            24th Mar 2006
          </div>
          <div
            className="css-isx2hi"
          >
            Falcon 1
          </div>
        </td>
      </tr>
      <tr
        className="css-1k3981y"
      >
        <td
          className="css-8hb6fg"
        >
          #2
        </td>
        <td
          className="css-o3rh2l"
        >
          DemoSat
        </td>
        <td
          className="css-1hyzgfm"
        >
          <div
            className="css-1wzm5rf"
          >
            21st Mar 2007
          </div>
          <div
            className="css-isx2hi"
          >
            Falcon 1
          </div>
        </td>
      </tr>
      <tr
        className="css-1k3981y"
      >
        <td
          className="css-8hb6fg"
        >
          #3
        </td>
        <td
          className="css-o3rh2l"
        >
          Trailblazer
        </td>
        <td
          className="css-1hyzgfm"
        >
          <div
            className="css-1wzm5rf"
          >
            3rd Aug 2008
          </div>
          <div
            className="css-isx2hi"
          >
            Falcon 1
          </div>
        </td>
      </tr>
      <tr
        className="css-1k3981y"
      >
        <td
          className="css-8hb6fg"
        >
          #4
        </td>
        <td
          className="css-o3rh2l"
        >
          RatSat
        </td>
        <td
          className="css-1hyzgfm"
        >
          <div
            className="css-1wzm5rf"
          >
            29th Sep 2008
          </div>
          <div
            className="css-isx2hi"
          >
            Falcon 1
          </div>
        </td>
      </tr>
      <tr
        className="css-1k3981y"
      >
        <td
          className="css-8hb6fg"
        >
          #5
        </td>
        <td
          className="css-o3rh2l"
        >
          RazakSat
        </td>
        <td
          className="css-1hyzgfm"
        >
          <div
            className="css-1wzm5rf"
          >
            13th Jul 2009
          </div>
          <div
            className="css-isx2hi"
          >
            Falcon 1
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`LaunchTable Snapshots Renders with fullest props 1`] = `
<div>
  <div
    className="css-m5epdm"
  >
    <div
      className="css-17w8ll9"
    >
      <div
        className="css-ovst58"
      >
        <select
          className="css-1bjzmyt"
          id="launch-year-filter-select"
          onChange={[Function]}
          value="2008"
        >
          <option
            value="Filter by Year"
          >
            Filter by Year
          </option>
          <option
            value="2006"
          >
            2006
          </option>
          <option
            value="2007"
          >
            2007
          </option>
          <option
            value="2008"
          >
            2008
          </option>
          <option
            value="2009"
          >
            2009
          </option>
        </select>
        <img
          alt="Chevron down"
          height="5px"
          srcSet="
			select@3x.png 3x,
			select@2x.png 2x,
			select@3x.png 1x,
		"
        />
      </div>
      <button
        className="css-4otj2c"
        onClick={[MockFunction]}
        type="button"
      >
        <span>
          Sort Descending
        </span>
        <img
          alt="Rocket launching"
          height="18px"
          srcSet="
      sort@2x.png 3x,
      sort.png 2x,
      sort.png 1x,
    "
        />
      </button>
    </div>
  </div>
  <table
    className="css-yn959"
  >
    <tbody>
      <tr
        className="css-1k3981y"
      >
        <td
          className="css-8hb6fg"
        >
          #3
        </td>
        <td
          className="css-o3rh2l"
        >
          Trailblazer
        </td>
        <td
          className="css-1hyzgfm"
        >
          <div
            className="css-1wzm5rf"
          >
            3rd Aug 2008
          </div>
          <div
            className="css-isx2hi"
          >
            Falcon 1
          </div>
        </td>
      </tr>
      <tr
        className="css-1k3981y"
      >
        <td
          className="css-8hb6fg"
        >
          #4
        </td>
        <td
          className="css-o3rh2l"
        >
          RatSat
        </td>
        <td
          className="css-1hyzgfm"
        >
          <div
            className="css-1wzm5rf"
          >
            29th Sep 2008
          </div>
          <div
            className="css-isx2hi"
          >
            Falcon 1
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
